tests/handlers/test_federation_event.py

```diff
@@ -848,3 +868,442 @@ async def get_room_state(
+            get_room_state_ids
+        )
+        self.mock_federation_transport_client.get_room_state.side_effect = (
+            get_room_state
+        )
+
+        self.mock_federation_transport_client.get_event.side_effect = get_event
+
+        # create the room
+        room_creator = self.appservice.sender
+        room_id = self.helper.create_room_as(
+            room_creator=self.appservice.sender, tok=self.appservice.token
+        )
+        room_version = self.get_success(main_store.get_room_version(room_id))
+
+        event_before = self.get_success(
+            inject_event(
+                self.hs,
+                room_id=room_id,
+                sender=room_creator,
+                type=EventTypes.Message,
+                content={"body": "eventBefore0", "msgtype": "m.text"},
+            )
+        )
+        _add_to_known_event_list(event_before)
+
+        event_after = self.get_success(
+            inject_event(
+                self.hs,
+                room_id=room_id,
+                sender=room_creator,
+                type=EventTypes.Message,
+                content={"body": "eventAfter0", "msgtype": "m.text"},
+            )
+        )
+        _add_to_known_event_list(event_after)
+
+        state_map = self.get_success(
+            state_storage_controller.get_state_for_event(event_before.event_id)
+        )
+
+        room_create_event = state_map.get((EventTypes.Create, ""))
+        pl_event = state_map.get((EventTypes.PowerLevels, ""))
+        as_membership_event = state_map.get((EventTypes.Member, room_creator))
+        assert room_create_event is not None
+        assert pl_event is not None
+        assert as_membership_event is not None
+
+        for state_event in state_map.values():
+            _add_to_known_event_list(state_event)
+
+        # This should be the successor of the event we want to insert next to
+        # (the successor of event_before is event_after).
+        inherited_depth = event_after.depth
+
+        historical_base_auth_event_ids = [
+            room_create_event.event_id,
+            pl_event.event_id,
+        ]
+        historical_state_events = list(state_map.values())
+        historical_state_event_ids = [
+            state_event.event_id for state_event in historical_state_events
+        ]
+
+        maria_mxid = "@maria:test"
+        maria_membership_event, _ = self.get_success(
+            create_event(
+                self.hs,
+                room_id=room_id,
+                sender=maria_mxid,
+                state_key=maria_mxid,
+                type=EventTypes.Member,
+                content={
+                    "membership": "join",
+                },
+                # It all works when I add a prev_event for the floating
+                # insertion event but the event no longer floats.
+                # It's able to resolve state at the prev_events though.
+                prev_event_ids=[event_before.event_id],
```

Comment:
This can work without connecting it to `event_before` but then we rely on `maria_membership_event` not being gossiped about during backfill because if it is, then it will be rejected and we can't use a rejected event to auth the following historical events.

```
_process_pulled_event
_process_pulled_event
_compute_event_context_with_maybe_missing_prevs
compute_event_context
```

The reason it works when it's not gossiped about is that `_compute_event_context_with_maybe_missing_prevs` fills in the `state_ids_before_event` and resolves the state magically for us without rejecting.

https://github.com/matrix-org/synapse/blob/5f659d4a88e602ca8519984808dcf4df036c781b/synapse/handlers/federation_event.py#L971-L1037

I have a suspicion that we will also stop doing sort of thing since we also removed auth event resolving in https://github.com/matrix-org/synapse/pull/12943